**Импорт всех необходимых библиотек**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option("display.max_rows", 20)
pd.set_option("display.max_columns", 20)
pd.set_option("display.precision", 4)
pd.set_option("plotting.backend", "matplotlib")

# 1. Исследовательский анализ данных (exploratory data analysis - EDA)

## 1.1 Словесное описание признаков

<b><p>data.csv</p></b>
<ul>
<li>id: id транзакции </li>
<li>timestamp: дата продажи (транзакции) </li>
<li>full_sq: общая площадь  </li>
<li>life_sq: жилая площадь  </li>
<li>floor: этаж  </li>
<li>max_floor: количество этажей в здании </li>
<li>material: материал, из которого изговолены стены  </li>
<li>build_year: год строительства </li>
<li>num_room: количество жилых комнат </li>
<li>kitch_sq: площадь кухни </li>
<li>full_all: количество населения в регионе </li>
<li> state: жилищные условия </li>
<li>sub_area: название территории </li>
<li>price_doc: цена квартиры (целевая переменная) </li>
</ul>

<b><p>macro.csv</p></b>
<ul>
<li>timestamp : дата, на которую актуальны макроэкономические показатели </li>
<li>salary : средняя зарплата в регионе </li>
<li>fixed_basket: стоимость потребительской корзины </li>
<li>rent_price_3room_eco: стоимость аренды 3-х комнатного жилья эконом-класса </li>
<li>rent_price_2room_eco: стоимость аренды 2-х комнатного жилья эконом-класса </li>
<li>rent_price_1room_eco: стоимость аренды 1-но комнатного жилья эконом-класса </li>
<li>average_life_exp: средняя продолжительность жизни в регионе </li>
</ul>

## 1.2 Загрузка данных общее описание набора данных

In [2]:
PATH_base = "https://raw.githubusercontent.com/aksenov7/Kaggle_competition_group/master/data.csv"
PATH_add = "https://raw.githubusercontent.com/aksenov7/Kaggle_competition_group/master/data_macro.csv"
df = pd.read_csv(PATH_base)
macro =  pd.read_csv(PATH_add)

### 1.2.1 Базовый датасет

Вывести пример данных (первые строки и случайные строки)

In [6]:
df.head(15)

id   timestamp  full_sq  life_sq  floor  state  max_floor  material  \
0   8059  2013-05-21       11     11.0    2.0    3.0        5.0       2.0   
1   8138  2013-05-25       53     30.0   10.0    3.0       16.0       1.0   
2   8156  2013-05-27       77     41.0    2.0    1.0       17.0       6.0   
3   8157  2013-05-27       45     27.0    6.0    3.0        9.0       1.0   
4   8178  2013-05-28       38     20.0   15.0    NaN       16.0       1.0   
5   8189  2013-05-28       63      NaN   17.0    1.0       17.0       4.0   
6   8258  2013-05-30       74     46.0   12.0    3.0       24.0       1.0   
7   8271  2013-05-31       93     93.0    3.0    1.0        1.0       1.0   
8   8285  2013-06-01       51     30.0    7.0    3.0       17.0       1.0   
9   8290  2013-06-03       77     50.0    3.0    2.0        5.0       2.0   
10  8296  2013-06-03       38     19.0   17.0    3.0       17.0       1.0   
11  8353  2013-06-06       43     29.0    3.0    3.0        5.0       2.0   
12  8365  2013-06-06       54     31.0    7.0    3.0       10.0       1.0   
13  8370  2013-06-06       44     28.0    1.0    2.0        9.0       1.0   
14  8377  2013-06-07       32     18.0    7.0    NaN        9.0       1.0   

   build_year  num_room  kitch_sq  full_all                  sub_area  \
0      1907.0       1.0      12.0     75377                 Hamovniki   
1      1980.0       2.0       8.0     68630                 Lianozovo   
2      2014.0       3.0      12.0      9553   Poselenie Voskresenskoe   
3      1970.0       2.0       6.0     78616           Severnoe Butovo   
4      1982.0       1.0       8.0    112804            Filevskij Park   
5      2013.0       2.0       1.0    247469                Nekrasovka   
6      2004.0       3.0       9.0     27992        Juzhnoe Medvedkovo   
7      2013.0       3.0       1.0      5740    Poselenie Pervomajskoe   
8      2003.0       2.0       9.0    125111                  Solncevo   
9      1957.0       3.0       8.0   1100773                 Ajeroport   
10     1986.0       1.0       8.0    129207  Orehovo-Borisovo Juzhnoe   
11     1960.0       2.0       5.0    115352                  Nagornoe   
12     1995.0       2.0       9.0    165727                   Mar'ino   
13     1979.0       2.0       6.0    156377                  Strogino   
14     1975.0       1.0       5.0    125354       Chertanovo Severnoe   

    price_doc  
0     2750000  
1     9000000  
2     7011550  
3     7100000  
4     6450000  
5     5903808  
6    12100000  
7     5427640  
8     7700000  
9    11700000  
10    1600000  
11    6400000  
12    9100000  
13    7200000  
14    5100000

In [7]:
df.sample(15)

id   timestamp  full_sq  life_sq  floor  state  max_floor  material  \
4041   14542  2014-02-04       30     16.0    6.0    3.0        9.0       2.0   
13179  25799  2014-12-04       57     37.0    2.0    2.0        5.0       2.0   
12880  25438  2014-11-27       37     22.0    1.0    2.0       12.0       5.0   
12566  25041  2014-11-20       31     19.0    4.0    2.0        4.0       2.0   
17575   3784  2012-09-22       79     41.0    6.0    NaN        NaN       NaN   
6972   18095  2014-05-05       44     27.0    5.0    2.0        5.0       2.0   
6586   17622  2014-04-19       31     21.0    8.0    2.0        9.0       1.0   
756    10485  2013-09-13       46     30.0    2.0    3.0       12.0       1.0   
15989  29321  2015-04-22       48     27.0   10.0    2.0       10.0       6.0   
4747   15387  2014-02-25       32     20.0    5.0    2.0        5.0       1.0   
8484   19943  2014-06-21       78     75.0    6.0    1.0       17.0       1.0   
14918  27976  2015-02-18       58     43.0    3.0    3.0        9.0       1.0   
6224   17200  2014-04-09       34     34.0    1.0    2.0       16.0       5.0   
13450  26160  2014-12-09       32     18.0    8.0    3.0        9.0       1.0   
7793   19097  2014-05-29       57     35.0    7.0    3.0       20.0       1.0   

      build_year  num_room  kitch_sq  full_all                  sub_area  \
4041      1971.0       1.0       5.0   1318695          Vyhino-Zhulebino   
13179     1958.0       2.0       8.0     36154                    Perovo   
12880     1963.0       2.0       5.0    153248                  Kotlovka   
12566     1959.0       1.0       5.0   1100773                 Ajeroport   
17575        NaN       NaN       NaN      9553   Poselenie Voskresenskoe   
6972      1964.0       2.0       5.0    122862                 Izmajlovo   
6586      1966.0       1.0       5.0     76156        Nagatino-Sadovniki   
756       1980.0       2.0       6.0     53786          Severnoe Tushino   
15989     2008.0       2.0       8.0    221709                  Krjukovo   
4747      1966.0       1.0       5.0    145576                  Caricyno   
8484      2014.0       3.0       1.0    247469                Nekrasovka   
14918     1974.0       3.0       6.0    129207  Orehovo-Borisovo Juzhnoe   
6224      1984.0       1.0       1.0    142462                Mozhajskoe   
13450     1975.0       1.0      10.0     61396                  Otradnoe   
7793      1973.0       2.0       7.0     73148                Vojkovskoe   

       price_doc  
4041     5500000  
13179    4000000  
12880    6600000  
12566    6150000  
17575    6780400  
6972     1000000  
6586     6000000  
756      7300000  
15989    6450000  
4747     4800000  
8484     7722400  
14918    8500000  
6224     5500000  
13450    6000000  
7793    13000000

Размер набора данных

In [8]:
len(df)

18866

Описание типов данных по признакам

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18866 entries, 0 to 18865
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          18866 non-null  int64  
 1   timestamp   18866 non-null  object 
 2   full_sq     18866 non-null  int64  
 3   life_sq     17291 non-null  float64
 4   floor       18838 non-null  float64
 5   state       15505 non-null  float64
 6   max_floor   17445 non-null  float64
 7   material    17445 non-null  float64
 8   build_year  16866 non-null  object 
 9   num_room    17445 non-null  float64
 10  kitch_sq    17445 non-null  float64
 11  full_all    18866 non-null  int64  
 12  sub_area    18866 non-null  object 
 13  price_doc   18866 non-null  int64  
dtypes: float64(7), int64(4), object(3)
memory usage: 2.0+ MB


Базовые статистики по признакам

In [12]:
df.describe(include='all')

id   timestamp     full_sq     life_sq       floor  \
count   18866.0000       18866  18866.0000  17291.0000  18838.0000   
unique         NaN        1074         NaN         NaN         NaN   
top            NaN  2014-12-16         NaN         NaN         NaN   
freq           NaN         133         NaN         NaN         NaN   
mean    18799.7886         NaN     53.9409     34.2592      7.2937   
std      7082.4287         NaN     44.1162     59.9446      5.1912   
min         9.0000         NaN      0.0000      0.0000      0.0000   
25%     13513.2500         NaN     38.0000     20.0000      3.0000   
50%     19043.5000         NaN     49.0000     30.0000      6.0000   
75%     24748.7500         NaN     63.0000     43.0000     10.0000   
max     30473.0000         NaN   5326.0000   7478.0000     77.0000   

             state   max_floor    material build_year    num_room    kitch_sq  \
count   15505.0000  17445.0000  17445.0000      16866  17445.0000  17445.0000   
unique         NaN         NaN         NaN        120         NaN         NaN   
top            NaN         NaN         NaN     2014.0         NaN         NaN   
freq           NaN         NaN         NaN        919         NaN         NaN   
mean        2.2013     12.4233      1.8901        NaN      1.9348      7.2064   
std         0.8528      6.2798      1.5061        NaN      0.8642     26.9816   
min         1.0000      0.0000      1.0000        NaN      0.0000      0.0000   
25%         2.0000      9.0000      1.0000        NaN      1.0000      5.0000   
50%         2.0000     12.0000      1.0000        NaN      2.0000      7.0000   
75%         3.0000     17.0000      2.0000        NaN      3.0000      9.0000   
max        33.0000     57.0000      6.0000        NaN     19.0000   2014.0000   

          full_all    sub_area   price_doc  
count   1.8866e+04       18866  1.8866e+04  
unique         NaN         146         NaN  
top            NaN  Nekrasovka         NaN  
freq           NaN         620         NaN  
mean    1.5972e+05         NaN  7.5824e+06  
std     3.0435e+05         NaN  4.9838e+06  
min     2.5460e+03         NaN  1.0000e+05  
25%     4.8439e+04         NaN  5.1000e+06  
50%     8.7713e+04         NaN  6.6855e+06  
75%     1.2535e+05         NaN  8.9000e+06  
max     1.7167e+06         NaN  9.5122e+07

Выводы текстом, что вы можете сказать по каждому признаку, на основе базовых статистик

1. Данные о продажах собраны за ~3 года (1074 уникальных дат), больше всего сделок (133) было совершено 16 декабря 2014 года
2. Средняя площадь проданных квартир составила ~54 кв. м., вызывает подозрения максимальная общая площадь: она меньше максимальной жилой площади, что невозможно.
   Также, очевидно, что в данных есть пропуски, так как минимальная площадь равна 0.
   Учитывая вышесказанное, делать какие-либо выводы на основании серднего значения и стандартного отклонения площади нельзя, так как в статистике явно присутсвует шум
   Также, подозрительно, что все данные в столбце целые (тип данных стобца опеределился как int64)
3. Аналогичные замечания справедливы и для жилой площади.
   Дополнительно можно заметить, что по ~1500 сделок отсутсвует информация по жилой площади
4. Большинство проданных квартир находятся на 2-12 этажах (среднее 7 +/- ст. отклонение 5), хотя среднему значению и стандартному отклонению и тут доверять нельзя:
   Минимальный этаж равен 0, а максимальный этаж больше максимальной этажности здания, что снова указывает на наличие шума
   Не хватает информации об этаже по 28 сделкам
   Подозрительно, что в данных могут присутствовать нецелочисленные значения (тип данных столбца определился как float64)
5. Жилищные условия, по идее, представляют собой оценки от 1 до 5 (если судить по распределению по квартилям), однако максимальное значение равно 33, что, скорее всего, ошибка при заполнении.
   Скорее всего, существует справочник Балл-Описание, добавление которого повысит интерпретируемость информации
   Оценки могут быть нецелочисленными (тип данных столбца определился как float64)
   По ~3300 сделок информация отсутсвует
6. По этажности здания комментарии аналогичны комментариям по этажу.
   Пропущена информация по ~1500 сделок
   Подозрительно, что в данных могут присутствовать нецелочисленные значения (тип данных столбца определился как float64)
7. Данные по материалу выглядят как перечень ID материлов, нужен справочник ID-Название, чтобы повысить интерпретируемость информации
   Всего в статистике представлено 6 различных материлов, наиболее популярным является материал с ID 1: стены как минимум половины всех проданных квартир сделаны из него.
   Среднее значение 1.9 и 75-й перцентиль, равный 2, указывают на то, что вторым по популярности является материал с ID 2
   По ~1500 сделок нет информации по материалу
   Подозрительно, что в данных могут присутствовать нецелочисленные значения (тип данных столбца определился как float64)
8. В данных, скорее всего, представлены в основном новостройки (наиболее популярый год постройки 2014 с 919 продажами).
   Однако присутвуют и дома дореволюционного периода (120 уникальных годов постройки).
   Не хватает информации о годе постройки по ~2000 проданных квартир
   Тип данных столбца (object) указывает на наличие текстовых записей
9. В основном продавались двухкомнатные квартиры (среднее значение 1.9)
   Большинство квартир имело от 1 до 3 жилых комнат (среднее +/- стандартное отклонение, распределение по квартилям)
   Минимальное значение 0 и максимальное 19 вызывают подозрения о наличии шума и пропусков, но их, скорее всего, немного, так как остальная статистика выглядит правдоподобно
   В данных пропущено ~1500 записей о количестве комнат
   Подозрительно, что в данных могут присутствовать нецелочисленные значения (тип данных столбца определился как float64)
10. По площади кухни комментарии аналогичны остальным параметрам площади
    Максимальная площадь 2014 кв. м., возможно, является годом постройки, ошибочно занесенным в поле "Площадь кухни" при заполнении
    Информации не хвататет по ~1500 сделок
11. Население территорий в среднем составляет ~160 тыс. жителей
    Самая густонаселенная территория имеет ~1.7 млн жителей
    На территории с минимальным населением проживает всего ~2.5 тыс. жителей
    Население территорий распределено сильно неравномерно: среднее выше 75-го перцентиля, максимум почти в 11 раз больше среднего и почти в 14 больше 75-го перцентиля
12. Больше всего квартир (620) было продано на территории Nekrasovka, всего территорий в данных представлено 146
13. Средняя цена сделки составила ~7.5 млн рублей, большая часть сделок не превышала 9 млн рублей
    Самая дешевая квартира стоила 100 тыс. рублей, что больше похоже на стоимость кв. м., чем на стоимость квартиры (норматив стоимости метра от Минстроя на 2H22 составляет 83 тыс. руб.)
    Стоимость указана без копеек (тип данных стобца опеределился как int64)

Какие признаки вы считаете полезными для предсказания цены квартиры, а какие по вашему мнению можно убрать. Почему?

Если не учитывать макроэкономические показатели и опираться исключительно на данные в рассматриваемой статистике, то полезными будут:
- Общая площадь (основной потребительский атрибут квартиры)
- Количество жилых комнат (характеризует "сценарий использования" квартиры - для студента/холостяка, для семьи с двумя детьми и проч.)
- Отношение жилой площади к количеству жилых комнат (характеризует качество квартиры, в паре с общей площадью позволяет косвенно учитывать площадь кухни, санузла и проч.)
- Отношение этажа к этажности здания (чем выше, тем дороже, но просто этажа не достаточно: 5 этаж в пятиэтажке и в 17-этажном здании будут стоить разных денег)
- Жилищные условия (зависит от используемых параметров оценки, скорее всего характеризует качество квартиры, здания и района)
- Территория (косвенно объединяет в себе все атрибуты территории: население, покупательная способность, объем спроса и предложения на рынке, без привязки к макроэкономическим показателям послужит прокси для них)

Убрать можно:
- Дату сделки (сама по себе никакой информации для предсказания не несет, может быть нужна только для связки с дополнительными данными)
- Материал стен (скорее всего отражен в жилищных условиях, так как влияет на шумоизоляцию и температуру в квартире)
- Год постройки (скорее всего отражен в жилищных условиях, так как влияет на степень износа и общедомовую инфраструктуру: лифт, мусоропровод и проч.)
- Площадь кухни (зависит от обшщей и жилой площади, не добавит значимой информации для предсказания)
- Население территории (является одним из атрибутов территории, без привязки к макроэкономическим показателям по ней дополнительной информации для предсказания не несет)
  (при попытке использования придется предсказывать численность населения территории, чтобы от нее предсказать уже стоимость квартиры, а если принять показатель статичным, то для предсказания хватит самого названия территории)

Есть ли пропуски в данных? В каких столбцах? Какой процент по каждому столбцу?

In [15]:
df.isna().sum() / len(df) * 100

id             0.0000
timestamp      0.0000
full_sq        0.0000
life_sq        8.3484
floor          0.1484
state         17.8151
max_floor      7.5321
material       7.5321
build_year    10.6011
num_room       7.5321
kitch_sq       7.5321
full_all       0.0000
sub_area       0.0000
price_doc      0.0000
dtype: float64

Есть ли аномальные данные в стллбцах? Если да, то укажите на них и объясните, почему считаете аномальными

Описано в текстовых выводах

### 1.2.2 Макро показатели

Вывести пример данных (первые строки и случайные строки)

In [16]:
macro.head(15)

timestamp   salary  fixed_basket  rent_price_3room_eco  \
0   2010-01-01  38410.5      11443.63                   NaN   
1   2010-01-02  38410.5      11443.63                   NaN   
2   2010-01-03  38410.5      11443.63                   NaN   
3   2010-01-04  38410.5      11443.63                   NaN   
4   2010-01-05  38410.5      11443.63                   NaN   
5   2010-01-06  38410.5      11443.63                   NaN   
6   2010-01-07  38410.5      11443.63                   NaN   
7   2010-01-08  38410.5      11443.63                   NaN   
8   2010-01-09  38410.5      11443.63                   NaN   
9   2010-01-10  38410.5      11443.63                   NaN   
10  2010-01-11  38410.5      11443.63                   NaN   
11  2010-01-12  38410.5      11443.63                   NaN   
12  2010-01-13  38410.5      11443.63                   NaN   
13  2010-01-14  38410.5      11443.63                   NaN   
14  2010-01-15  38410.5      11443.63                   NaN   

    rent_price_2room_eco  rent_price_1room_eco  average_life_exp  
0                    NaN                   NaN              74.2  
1                    NaN                   NaN              74.2  
2                    NaN                   NaN              74.2  
3                    NaN                   NaN              74.2  
4                    NaN                   NaN              74.2  
5                    NaN                   NaN              74.2  
6                    NaN                   NaN              74.2  
7                    NaN                   NaN              74.2  
8                    NaN                   NaN              74.2  
9                    NaN                   NaN              74.2  
10                   NaN                   NaN              74.2  
11                   NaN                   NaN              74.2  
12                   NaN                   NaN              74.2  
13                   NaN                   NaN              74.2  
14                   NaN                   NaN              74.2

In [17]:
macro.sample(15)

timestamp   salary  fixed_basket  rent_price_3room_eco  \
1474  2014-01-14  61208.0      15608.27                 47.04   
795   2012-03-06  48830.4      13243.77                 45.31   
2070  2015-09-02  64310.0      19088.53                 50.07   
2164  2015-12-05  64310.0      19274.42                 49.42   
1136  2013-02-10  55485.2      14377.54                 47.90   
1201  2013-04-16  55485.2      14739.33                 47.09   
2456  2016-09-22      NaN      20354.78                 45.71   
862   2012-05-12  48830.4      13350.32                 44.57   
1612  2014-06-01  61208.0      16295.00                 48.70   
2223  2016-02-02      NaN      19544.19                 46.89   
2388  2016-07-16      NaN      20422.50                 46.79   
328   2010-11-25  38410.5      12094.33                 42.07   
2157  2015-11-28  64310.0      19213.32                 49.62   
250   2010-09-08  38410.5      11920.13                   NaN   
1388  2013-10-20  55485.2      15285.65                 48.28   

      rent_price_2room_eco  rent_price_1room_eco  average_life_exp  
1474                 41.44                 33.49             76.70  
795                  38.45                 31.64             75.74  
2070                 39.06                 30.69             76.77  
2164                 40.09                 30.37             76.77  
1136                  0.10                 32.62             76.37  
1201                 41.30                 32.61             76.37  
2456                 38.40                 29.78               NaN  
862                  37.96                 30.67             75.74  
1612                 41.18                 33.06             76.70  
2223                 38.11                 29.85               NaN  
2388                 37.74                 29.83               NaN  
328                  35.20                 28.88             74.20  
2157                 39.48                 30.48             76.77  
250                    NaN                   NaN             74.20  
1388                 41.96                 34.52             76.37

Размер набора данных

In [18]:
len(macro)

2484

Описание типов данных по признакам

In [19]:
macro.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2484 entries, 0 to 2483
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   timestamp             2484 non-null   object 
 1   salary                2191 non-null   float64
 2   fixed_basket          2484 non-null   float64
 3   rent_price_3room_eco  2211 non-null   float64
 4   rent_price_2room_eco  2211 non-null   float64
 5   rent_price_1room_eco  2211 non-null   float64
 6   average_life_exp      2191 non-null   float64
dtypes: float64(6), object(1)
memory usage: 136.0+ KB


Базовые статистики по признакам

In [20]:
macro.describe(include='all')

timestamp      salary  fixed_basket  rent_price_3room_eco  \
count         2484   2191.0000     2484.0000             2211.0000   
unique        2484         NaN           NaN                   NaN   
top     2010-01-01         NaN           NaN                   NaN   
freq             1         NaN           NaN                   NaN   
mean           NaN  52188.9331    15331.4980               47.2997   
std            NaN   9068.5590     2789.1363                3.0702   
min            NaN  38410.5000    11443.6300               37.5400   
25%            NaN  44898.7000    12992.4400               45.5300   
50%            NaN  48830.4000    14983.9200               47.3500   
75%            NaN  61208.0000    18295.0700               49.3500   
max            NaN  64310.0000    20422.5000               53.2100   

        rent_price_2room_eco  rent_price_1room_eco  average_life_exp  
count              2211.0000             2211.0000         2191.0000  
unique                   NaN                   NaN               NaN  
top                      NaN                   NaN               NaN  
freq                     NaN                   NaN               NaN  
mean                 38.8954               31.0024           75.9282  
std                   5.0352                3.9841            0.8699  
min                   0.1000                2.3100           74.2000  
25%                  37.9300               29.7800           75.7400  
50%                  39.4700               31.1100           75.7900  
75%                  41.3000               33.0600           76.7000  
max                  43.8500               35.3700           76.7700

Выводы текстом, что вы можете сказать по каждому признаку, на основе базовых статистик

1. В статистике представлены данные за ~7 лет (2484 уникальных даты)
2. Средняя зарплата в регионе за описываемый период составила ~52 тыс. руб. с разбросом от ~38 тыс. руб. до ~64 тыс. руб.
   Подозрительных / аномальных значений в базовой статистике нет
   Не хвататет данных по ~300 наблюдениям
3. Средняя стоимость потребительской корзины составила ~15 тыс. руб. с разбросом от ~11.5 тыс. руб. до ~20.5 тыс. руб.
   Подозрительных / аномальных значений в базовой статистике нет
4. Средняя стоимость аренды 3-х комнатной квартиры составила ~47 тыс. руб. с разбросом от ~37 тыс. руб. до ~53 тыс. руб.
   Подозрительных / аномальных значений в базовой статистике нет
   Пропущено ~1200 записей
5. Средняя стоимость аренды 2-х комнатной квартиры составила ~39 тыс. руб. с разбросом от 100 руб. до ~44 тыс. руб.
   Минимальная стоимость в 100 рублей явно указывает на наличие шума в данных, но его, скорее всего, немного, так как среднее значение укладывается в общую логику удорожания стоимости аренды:
   Средняя стоимость 2-х комнатной на ~25% выше средней стоимости однокомнатной, средняя стоимость 3-х комнатной на ~22% выше средней стоимости 2-х комнатной
   Пропущено ~1200 записей
6. Средняя стоимость аренды однокомнатной квартиры составила ~31 тыс. руб. с разбросом от ~30 тыс. руб. до ~35 тыс. руб.
   Подозрительных / аномальных значений в базовой статистике нет
   Пропущено ~1200 записей
7. Средняя продолжительность жизни составила ~76 лет с разбросом от ~74 лет до ~77 лет
   Подозрительных / аномальных значений в базовой статистике нет
   Пропущено ~300 записей

Количество пропусков указывает на наличие двух дополнительных источников, данные из которых притягивались к информации о стоимости потребительской корзины:
- Динамика зарплат и средней продожительности жизни
- Динамика стоимости аренды квартир

Какие признаки вы считаете полезными для предсказания цены квартиры, а какие по вашему мнению можно убрать. Почему?

Полезные признаки:
- Разница зарплаты и стоимости корзины (характеризует количество "свободных" денег у среднего жителя на оплату ипотечного платежа)
  (С другой стороны, учитывая, что данные представлены не по территориям, может оказаться бесполезным, так как абсолютная "платежеспособность" территории будет определяться ее населением, а относительная будет для всех территорий одинаковой)
- Стоимости аренды квартиры (характеризует стоимость альтернативы покупки квартиры и, как следствие, средний размер ипотечного платежа => средняя рыночная стоимость квартиры, если принять
  условия ипотеки константой)

Можно убрать:
- Дату (сама по себе никакой информации для предсказания не несет, может быть нужна только для связки с дополнительными данными)
- Продолжительность жизни (с натяжкой можно предположить, что чем выше продолжительность жизни, тем больше одобрят ипотечных кредитов, но:
  1) Подавляющее большинство покупателей квартир моложе чем [продолжительность жизни - срок ипотеки]
  2) Сама по себе вероятность одобрения очень косвенно влияет на цену квартиры)

Есть ли пропуски в данных? В каких столбцах? Какой процент по каждому столбцу?

In [21]:
macro.isna().sum() / len(df) * 100

timestamp               0.0000
salary                  1.5531
fixed_basket            0.0000
rent_price_3room_eco    1.4470
rent_price_2room_eco    1.4470
rent_price_1room_eco    1.4470
average_life_exp        1.5531
dtype: float64

Есть ли аномальные данные в стллбцах? Если да, то укажите на них и объясните, почему считаете аномальными

Описано в текстовых выводах

## 1.3 Замените все ранее найденные ошибочные данные на None. Заполните все пропуски в данных: которые были и которые появились. Используйте как стратегии изученные на занятии, так и логику, которая вытекает из самих данных

### 1.3.1 Главный набор данных

In [28]:
# Сохраняем изначальный датафрейм
df_init = df
df.describe()

id     full_sq     life_sq       floor       state   max_floor  \
count  18866.0000  18866.0000  17291.0000  18838.0000  15505.0000  17445.0000   
mean   18799.7886     53.9409     34.2592      7.2937      2.2013     12.4233   
std     7082.4287     44.1162     59.9446      5.1912      0.8528      6.2798   
min        9.0000      0.0000      0.0000      0.0000      1.0000      0.0000   
25%    13513.2500     38.0000     20.0000      3.0000      2.0000      9.0000   
50%    19043.5000     49.0000     30.0000      6.0000      2.0000     12.0000   
75%    24748.7500     63.0000     43.0000     10.0000      3.0000     17.0000   
max    30473.0000   5326.0000   7478.0000     77.0000     33.0000     57.0000   

         material    num_room    kitch_sq    full_all   price_doc  
count  17445.0000  17445.0000  17445.0000  1.8866e+04  1.8866e+04  
mean       1.8901      1.9348      7.2064  1.5972e+05  7.5824e+06  
std        1.5061      0.8642     26.9816  3.0435e+05  4.9838e+06  
min        1.0000      0.0000      0.0000  2.5460e+03  1.0000e+05  
25%        1.0000      1.0000      5.0000  4.8439e+04  5.1000e+06  
50%        1.0000      2.0000      7.0000  8.7713e+04  6.6855e+06  
75%        2.0000      3.0000      9.0000  1.2535e+05  8.9000e+06  
max        6.0000     19.0000   2014.0000  1.7167e+06  9.5122e+07

In [32]:
# Оцениваем аномалии с площадями
#df[df['full_sq'] == 5326]['price_doc'] / df[df['full_sq'] == 5326]['full_sq']
df.sort_values(by=['full_sq'], ascending=False).head(15)

id   timestamp  full_sq  life_sq  floor  state  max_floor  material  \
18428   3530  2012-09-07     5326     22.0   13.0    NaN        NaN       NaN   
10813  22788  2014-09-23      637    637.0   18.0    1.0       19.0       4.0   
18598   2428  2012-06-07      412     17.0    6.0    NaN        NaN       NaN   
14772  27796  2015-02-09      407    407.0    4.0    1.0        4.0       4.0   
18510  16746  2014-03-31      303      NaN   32.0    NaN       32.0       1.0   
7322   18528  2014-05-16      275    232.0   11.0    1.0       37.0       6.0   
6294   17280  2014-04-11      220    176.0    8.0    1.0       12.0       4.0   
15211  28329  2015-03-12      220    144.0    3.0    3.0        5.0       2.0   
6660   17713  2014-04-22      219    118.0   31.0    1.0       48.0       4.0   
15005  28084  2015-02-26      219    101.0    2.0    1.0        2.0       4.0   
11683  23899  2014-10-23      218    174.0   10.0    3.0       12.0       4.0   
12671  25186  2014-11-24      216    121.0   27.0    4.0       30.0       4.0   
4544   15148  2014-02-18      215    134.0   10.0    1.0       10.0       2.0   
16232  29659  2015-05-12      211    162.0    5.0    3.0        5.0       2.0   
14521  27499  2015-01-24      209      NaN   37.0    1.0        1.0       1.0   

      build_year  num_room  kitch_sq  full_all              sub_area  \
18428        NaN       NaN       NaN   1716730  Birjulevo Vostochnoe   
10813     2016.0       2.0      10.0    116742              Tverskoe   
18598        NaN       NaN       NaN    112804        Filevskij Park   
14772     2015.0       1.0      10.0    112804        Filevskij Park   
18510        NaN       6.0       0.0     83844          Obruchevskoe   
7322      2009.0       5.0      26.0    166803              Shhukino   
6294      2007.0       2.0      12.0     73148            Vojkovskoe   
15211     1927.0       9.0      17.0     57999           Presnenskoe   
6660      2010.0       3.0       4.0    102726             Jakimanka   
15005     2014.0       6.0      16.0    102618        Juzhnoe Butovo   
11683     2007.0       2.0      13.0     73148            Vojkovskoe   
12671     2003.0       4.0      20.0    102726             Jakimanka   
4544      2003.0       6.0       7.0     37807               Savelki   
16232     1914.0       6.0      19.0     57999           Presnenskoe   
14521     2015.0       1.0       1.0    175518             Rostokino   

       price_doc  
18428    6868818  
10813    4725142  
18598    6117000  
14772    3134129  
18510   54212936  
7322    15919294  
6294    37325704  
15211   95122496  
6660    14526873  
15005   18000000  
11683   46191272  
12671   64000000  
4544    23000000  
16232   70000000  
14521   44682580

### 1.3.2 Набор с макропоказателями

## 1.4 Обогатите основной набор данных данными из макропоказателей и поместите в переменную `df_full`

## 1.5 Проверьте данные на наличие выбросов. По каждому столбцу. Напишите своё мнение: нужно ли в каждой из ситуаций обрабатывать выбросы, или можно оставить. Если нужно обработать, то примените один из изученных подходов, либо предложите свой

### 1.5.1 Главный набор данных

### 1.5.2 Набор с макропоказателями

## 1.6 Создайте не менее 5 новых признаков на основе существующих данных. Опишите текстом обоснование создания каждой. Признаки должны привносить некую новую информацию для понимания цены квартиры

## 1.7 Провести визуальный анализ всех признаков

Ко всем графикам писать выводы текстом. Что вы видите, почему построили именно такую диаграмму

### 1.7.1 Анализ признаков по отдельности (где нужно делать группировки, преобразования категориальных типов данных к числовым, смотреть в разрезе других категориальных признаков)

### 1.7.2 Анализ совместного влияния признаков и их влияния на целевой признак